azavea/pfb-network-connectivity
# Add S3 tile caching to Terraform config

## Overview

Support Tilegarden caching by creating an S3 bucket for cached tiles, configuring CloudFront to try the cache first, and updating the app containers to enable caching.

### Notes

* I had to make a few small adjustments to Tilegarden to get this up and running. Because of that, I've requested @KlaasH's review on this as well.

## Testing Instructions

 * In the VM, run a plan/apply cycle and confirm that only the task definitions get updated (to match new batch jobs that `plan` creates):

```console
vagrant@vagrant-ubuntu-trusty-64:~$ export GIT_COMMIT=a7daa2a
vagrant@vagrant-ubuntu-trusty-64:~$ export ENVIRONMENT=staging
vagrant@vagrant-ubuntu-trusty-64:~$ export PFB_AWS_ECR_ENDPOINT="950872791630.dkr.ecr.us-east-1.amazonaws.com"
vagrant@vagrant-ubuntu-trusty-64:~$ export PFB_SETTINGS_BUCKET=staging-pfb-config-us-east-1
vagrant@vagrant-ubuntu-trusty-64:~$ ./scripts/infra plan
...
vagrant@vagrant-ubuntu-trusty-64:~$ ./scripts/infra apply
```

* Visit the staging site at https://staging.pfb.azavea.com, select `All Places`, and choose a place to load up some tiles
* If the view has been visited before: you should see rapid tile display, and the network tab should show that the tiles are coming from `tiles.staging.pfb.azavea.com/` 
    * This behavior is visible on https://staging.pfb.azavea.com/#/places/65302256-225c-49db-be9b-19bb51a96bb9/ 
* If the view has not been visited before: Lambda should take a few seconds to generate the tiles, and the network tab should show that the tiles are coming from `tiles.staging.pfb.azavea.com/latest` (the non-cached endpoint)
    * Hard refreshing the page may not confirm that the tile has been written to the cache, since (I believe) CloudFront will cache the 302 redirect to `/latest` for some short amount of time. To test that the cache is getting written, you can do two things:
        1. Check the cache S3 bucket `staging-pfb-tile-cache-us-east-1` to confirm that the path to the tile you've requested exists in the cache
        2. Create a CloudFront invalidation for `/tiles/*` to clear the CloudFront cache, and then check your endpoint again to confirm that the tile got cached

Closes #634.
